gpu-miner.c

```diff
@@ -148,91 +131,86 @@ int main() {
+	if (ret != CL_SUCCESS) { printf("failed to set third kernel arg: \n"); exit(1); }
 	ret = clSetKernelArg(kernel, 3, sizeof(cl_mem), (void *)&nonceOutmobj);
-	if (ret != CL_SUCCESS) { printf("failed to set second kernel arg: \n"); return -1; }
+	if (ret != CL_SUCCESS) { printf("failed to set second kernel arg: \n"); exit(1); }
 	ret = clSetKernelArg(kernel, 4, sizeof(cl_mem), (void *)&nonceOutLockmobj);
-	if (ret != CL_SUCCESS) { printf("failed to set fourth kernel arg: \n"); return -1; }
+	if (ret != CL_SUCCESS) { printf("failed to set fourth kernel arg: \n"); exit(1); }
+	ret = clSetKernelArg(kernel, 5, sizeof(cl_mem), (void *)&iter_per_threadmobj);
+	if (ret != CL_SUCCESS) { printf("failed to set fourth kernel arg: \n"); exit(1); }
+
+	// Rough scan for 'optimal' thread count
+	double hash_rate, prev_hash_rate = 0;
+	global_item_size = 16;
+	iter_per_thread = 16 * 256;
+	while(1) {
+		global_item_size *= 2;
+		double temp = grindNonces(global_item_size, iter_per_thread);
+		while (temp == -1) {
+			// Repeat until no block is found
+			temp = grindNonces(global_item_size, iter_per_thread);
+			printf("\rMining at %.3f MH/s\t%u blocks mined", hash_rate, blocks_mined);
+			fflush(stdout);
+		}
+		hash_rate = temp;
+		printf("\rMining at %.3f MH/s\t%u blocks mined", hash_rate, blocks_mined);
+		fflush(stdout);
+		if (hash_rate < prev_hash_rate) {
+			break;
+		}
+		prev_hash_rate = hash_rate;
+	}
+	printf("\rRough search found %zd threads to be the best at %.3f MH/s\n", global_item_size/2, prev_hash_rate);
+	fflush(stdout);
+
+	// Now we know the optimal is betweem global_item_size and global_item_size / 2
+	// Scan intermediate 16 values and pick the highest
+	int dec = (global_item_size - global_item_size / 2) / 16;
+	double best_hash_rate = prev_hash_rate;
+	size_t best_item_size = global_item_size / 2;
+	for (i = 0; i < 16; i++) {
+		global_item_size -= dec;
+		double temp = grindNonces(global_item_size, iter_per_thread);
+		while (temp == -1) {
+			// Repeat until no block is found
+			temp = grindNonces(global_item_size, iter_per_thread);
+			printf("\rMining at %.3f MH/s\t%u blocks mined", hash_rate, blocks_mined);
+			fflush(stdout);
+		}
+		hash_rate = temp;
+		printf("\rMining at %.3f MH/s\t%u blocks mined", hash_rate, blocks_mined);
+		fflush(stdout);
+		if (hash_rate > best_hash_rate) {
+			best_hash_rate = hash_rate;
+			best_item_size = global_item_size;
+		}
+	}
+	global_item_size = best_item_size;
+	printf("\rFine search found %zd threads to be the best at %.3f MH/s\n", global_item_size, best_hash_rate);
+	fflush(stdout);
 
-	// Mine blocks until program is interrupted
-	// Each iteration of the loop should take 1-3 seconds
+	// Make each iteration take about 3 seconds
+	clock_t startTime = clock();
+	grindNonces(global_item_size, iter_per_thread);
+	double run_time_seconds = (double)(clock() - startTime) / CLOCKS_PER_SEC;
+	iter_per_thread *= 3 / run_time_seconds;
```

Comment:
interesting that you would tune this live.

1. are you sure it doesn't affect hashrate?
2. is the big worry that the clockrate will change and slow things down? Is there another worry.

Ultimately it seems like a good idea but ' 3 / run_time_seconds' seems weird. You end up with non-even iteration counts.